src/csexp_rpc/csexp_rpc.ml

```diff
@@ -203,6 +219,30 @@ module Session = struct
       debug res;
       res
 
+  external send : Unix.file_descr -> Bytes.t -> int -> int -> int = "dune_send"
+
+  let write = if Sys.linux then send else Unix.single_write
+
+  let rec csexp_write_loop fd out_buf token write_mutex =
+    Mutex.lock write_mutex;
+    if Io_buffer.flushed out_buf token then Mutex.unlock write_mutex
+    else
+      (* We always make sure to try and write the entire buffer.
+          This should minimize the amount of [write] calls we need
+          to do *)
+      let written =
+        let bytes = Io_buffer.bytes out_buf in
+        let pos = Io_buffer.pos out_buf in
+        let len = Io_buffer.length out_buf in
+        try write fd bytes pos len
+        with exn ->
+          Mutex.unlock write_mutex;
+          reraise exn
+      in
+      Io_buffer.read out_buf written;
+      Mutex.unlock write_mutex;
```

Comment:
I think it's better because it's allowing callers to refill the write buffer on every send. Less write calls and more mutex locks sounds like a better trade off. Anyway, the mutex will be gone soon so I suggest we don't dwell on it.